Update test guide - Closes #514 (#518)
LiskHQ/lisk-docs: modules/ROOT/pages/guides/test.adoc

Mona Bärenfänger <mona@lightcurve.io>
:source-linenums-option:
:highlightjs-theme: solarized_dark
:url_github_hello_world: https://github.com/LiskHQ/lisk-sdk-examples/tree/development/hello_world
:url_github_hello_world_tx: https://github.com/LiskHQ/lisk-sdk-examples/blob/development/hello_world/transactions/hello_transaction.js#L53
:url_github_hello_world_tx_test: https://github.com/LiskHQ/lisk-sdk-examples/blob/test-for-hello-world/hello_world/transactions/test/hello_transaction.test.js
:url_jest: https://jestjs.io/docs/en/getting-started
:url_unit_tests: https://en.wikipedia.org/wiki/Unit_testing
:url_wiki_stubs: https://en.wikipedia.org/wiki/Test_stub

:url_tutorials_transport_3: tutorials/transport2.adoc#register_packet

When the logic inside the custom transaction types becomes more complex, it becomes complicated to verify that the custom transaction logic is working as expected.

> Therefore it is recommended to write {url_unit_tests}[unit tests], that verify the logic of the transaction type.

Reviewer: ```suggestion
Therefore it is recommended to write {url_unit_tests}[unit tests], that verifies the logic of the transaction type.
```

Author: I think it was correct as it was

Reviewer: yes, you are right Mona, please leave it as it was.

To undo the transaction, the `undoAsset()` function will be called for each transaction inside of the blocks that need to be reverted.
====

> To explain how to write unit tests for custom transactions, we will use the example of the `applyAsset()` method of the {url_github_hello_world}[hello world app] .

Reviewer: ```suggestion
In order to explain how to write unit tests for custom transactions, the `applyAsset()` method of the {url_github_hello_world}[hello world app] example is used here.
```

> You can use it as template to write unit tests for your own custom transactions.

Reviewer: ```suggestion
It can be used as a template to write unit tests for your own custom transactions.
```

NOTE: To test the complete logic of a custom transaction, it is required to write unit tests for all methods that are implemented in the custom transactions: `validateAsset()`, `applyAsset()` and `undoAsset()`.

== Write a test script

> First, let's look at the `applyAsset()` function of the {url_github_hello_world_tx}[HelloTransaction].

Reviewer: ```suggestion
Firstly, it is recommended to view the `applyAsset()` function of the {url_github_hello_world_tx}[HelloTransaction].
```

> First, let's look at the `applyAsset()` function of the {url_github_hello_world_tx}[HelloTransaction].

Reviewer: ```suggestion
Firstly, it is recommended to view the `applyAsset()` function of the {url_github_hello_world_tx}[HelloTransaction].
```

Author: I would not say "it's recommended" here.
It's just normal to look at the method you want to test.

Reviewer: OK, then in this case we can leave it as it is.. :-)

.Snippet of undoAsset(), File: hello_world/transactions/hello_transaction.js
[source,js]
----
applyAsset(store) {
    const errors = [];
    const sender = store.account.get(this.senderId);
    if (sender.asset && sender.asset.hello) {
        errors.push(
            new TransactionError(
                'You cannot send a hello transaction multiple times',
                sender.asset.hello,
                '.asset.hello',
                this.asset.hello
            )
        );
    } else {
        const newObj = { ...sender, asset: { hello: this.asset.hello } };
        store.account.set(sender.address, newObj);
    }
    return errors;
}
----

> To test, that the `applyAsset()` function is working as expected, write a unit test with {url_jest}[jest]: {url_github_transport_test_register}[hello_transaction.test.js]:

Reviewer: ```suggestion
To test that the `applyAsset()` function is working as expected, write a unit test with {url_jest}[jest]: {url_github_transport_test_register}[hello_transaction.test.js]:
```

        // Arrange
        const asset = {
            hello: "my hello message",
        };

        const senderId = '16313739661670634666L';
        const senderPublicKey = 'c094ebee7ec0c50ebee32918655e089f6e1a604b83bcaa760293c61e0f18ab6f';
        const mockedSenderAccount = {
            address: '16313739661670634666L',
            publicKey: 'c094ebee7ec0c50ebee32918655e089f6e1a604b83bcaa760293c61e0f18ab6f',
            asset: {},
        };

        when(storeStub.account.get) <3>
            .calledWith(senderId)
            .mockReturnValue(mockedSenderAccount);

        // Act
        const tx = new HelloTransaction({ <4>
            senderPublicKey,
            asset,
            timestamp: utils.getTimeFromBlockchainEpoch(new Date()),
        });
        tx.applyAsset(storeStub); <5>

        // Assert
        expect(storeStub.account.set).toHaveBeenCalledWith( <6>
            mockedSenderAccount.address,
            {
                address: mockedSenderAccount.address,
                publicKey: 'c094ebee7ec0c50ebee32918655e089f6e1a604b83bcaa760293c61e0f18ab6f',
                asset,
            }
        );
    });

>     it('should reject the transaction, if the sender has already a hello string in their account.', async () => {

Reviewer: ```suggestion
    it('should reject the transaction, if the sender already has a hello string in their account.', async () => {
```

        const mockedSenderAccount = {
                address: '16313739661670634666L',
                publicKey: 'c094ebee7ec0c50ebee32918655e089f6e1a604b83bcaa760293c61e0f18ab6f',
                asset: {
                    hello: "Hello world!"
                }
            };
        const errors = [];
        errors.push(
            new TransactionError(
                'You cannot send a hello transaction multiple times',
                mockedSenderAccount.asset.hello,
                '.asset.hello',
                asset.hello
            )
        );

        when(storeStub.account.get)
            .calledWith(senderId)
            .mockReturnValue(mockedSenderAccount);

        // Act
        const tx = new HelloTransaction({
            senderPublicKey,
            asset,
            timestamp: utils.getTimeFromBlockchainEpoch(new Date()),
        });

        // Assert
        expect(tx.applyAsset(storeStub)).toMatchObject(errors);
    });
});
----

<1> The `get()` and `set()` functions of the `stateStore` are mocked by creating {url_wiki_stubs}[stubs] in the `beforeEach()` function.

> This allows replace the call with a custom return value.

Reviewer: ```suggestion
This allows the possibility to replace the call with a custom return value.
```

> This allows replace the call with a custom return value.

Reviewer: ```suggestion
This allows the user to have the opportunity to replace the call with a custom return value.
```

> Stubbing allows to replace the call of a function with a custom return value.

Reviewer: ```suggestion
Stubbing provides the opportunity to replace the call of a function with a custom return value.
```

> <2> Here starts the first test, with a short and precise description of what the test is about.

Reviewer: ```suggestion
<2> Now the first test can begin, with a short and precise description of the actual test itself.
```

> <3> When `storeStub.account.get` is called with `asset.senderId`, we replace the return value with the `mockedSenderAccount`.

Reviewer: ```suggestion
<3> When `storeStub.account.get` is called with `asset.senderId`,the return value is replaced with the `mockedSenderAccount`.
```

<4> A new transaction is created.
<5> The `applyAsset()` function of the transaction is called, and the previousely defined `storeStub` is passed to the `applyAsset()` function.
<6> The test passes if the function `storeStub.account.set()` is called with the expected parameters.

The first test verifies that the hello string is saved in the senders account assets.

> Therefore, we check if `storeStub.account.set()` got called with the right parameters:

Reviewer: ```suggestion
Therefore, a check is performed to ensure that the `storeStub.account.set()` was called with the correct parameters:
```

.Sender address
[source,js]
----
mockedSenderAccount.address,
----

and
.Updated sender account
[source,js]
----
{
    address: mockedSenderAccount.address,
    publicKey: 'c094ebee7ec0c50ebee32918655e089f6e1a604b83bcaa760293c61e0f18ab6f',
    asset,
}
----

> If the function got called with the expected parameters, we know that the sender account was updated correctly.

Reviewer: ```suggestion
If the function was called with the expected parameters, then this proves that the sender account was updated correctly.
```

> In the second test, we want to verify that the transaction gets rejected, in case the sender has already a hello string in their account.

Reviewer: ```suggestion
In the second test it is necessary to verify that the transaction is rejected, in the case whereby the sender already has a hello string in their account.
```

> Therefore, we check if the expected `TransactionError` is returned when `applyAsset()` is executed.

Reviewer: ```suggestion
Therefore, it is necessary to check if the expected `TransactionError` is returned when the `applyAsset()` is executed.
```
